overlord/devicestate/handlers_systems.go

```diff
@@ -0,0 +1,374 @@
+	}
+
+	observeSnapFileWrite := func(recoverySystemDir, where string) error {
+		if recoverySystemDir != systemDirectory {
+			return fmt.Errorf("internal error: unexpected recovery system path %q", recoverySystemDir)
+		}
+		// track all the files, both asserted shared snaps and private
+		// ones
+		return logNewSystemSnapFile(filepath.Join(recoverySystemDir, "snapd-new-file-log"), where)
+	}
+
+	db := assertstate.DB(st)
+	defer func() {
+		if err == nil {
+			return
+		}
+		if err := purgeNewSystemSnapFiles(filepath.Join(systemDirectory, "snapd-new-file-log")); err != nil {
+			logger.Noticef("when removing seed files: %v", err)
+		}
+		// this is ok, as before the change with this task was created,
+		// we checked that the system directory did not exist; it may
+		// exist now if one of the post-create steps failed, or the the
+		// task is being re-run after a reboot and creating a system
+		// failed
+		if err := os.RemoveAll(systemDirectory); err != nil && !os.IsNotExist(err) {
+			logger.Noticef("when removing recovery system %q: %v", label, err)
+		}
+		if err := boot.DropRecoverySystem(remodelCtx, label); err != nil {
+			logger.Noticef("when dropping the recovery system %q: %v", label, err)
+		}
+		// we could have reentered the task after a reboot, but the
+		// state was set up sufficiently such that the system was
+		// actually tried and ended up in the tried systems list, which
+		// we should reset now
+		st.Set("tried-systems", nil)
+	}()
+	// 1. prepare recovery system from remodel snaps (or current snaps)
+	_, err = createSystemForModelFromValidatedSnaps(model, label, db, infoGetter, observeSnapFileWrite)
+	if err != nil {
+		return fmt.Errorf("cannot create a recovery system with label %q for %v: %v", label, model.Model(), err)
+	}
+	logger.Debugf("recovery system dir: %v", systemDirectory)
+
+	// 2. keep track of the system in task state
+	if err := setTaskRecoverySystemSetup(t, setup); err != nil {
+		return fmt.Errorf("cannot record recovery system setup state: %v", err)
+	}
+	// 3. set up boot variables for tracking the tried system state
+	if err := boot.SetTryRecoverySystem(remodelCtx, label); err != nil {
+		// rollback?
+		return fmt.Errorf("cannot attempt booting into recovery system %q: %v", label, err)
+	}
+	// 4. and set up the next boot that that system
+	if err := boot.SetRecoveryBootSystemAndMode(remodelCtx, label, "recover"); err != nil {
+		return fmt.Errorf("cannot set device to boot into candidate system %q: %v", label, err)
+	}
+
+	// this task is done, further processing happens in finalize
+	t.SetStatus(state.DoneStatus)
+
+	logger.Noticef("restarting into candidate system %q", label)
+	m.state.RequestRestart(state.RestartSystemNow)
+	return nil
+}
+
+func (m *DeviceManager) undoCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	if release.OnClassic {
+		// TODO: this may need to be lifted in the future
+		return fmt.Errorf("cannot run update gadget assets task on a classic system")
```

Comment:
same, though this is definitely an internal error because we should have failed on the do path already